### PR TITLE
Sum_numbers first commit

### PR DESCRIPTION
**Pull request by Maab232001**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/sum_numbers.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+Team number:28
+Team name:MIT Alpha
+The author: Maab Mohamedkhair"""
+
+
+def sum_numbers(n):
+    
+    """ sum numbers 
+sum numbers is a recursive function return the sum of count down numbers 
+from n -> 1
```
Comment (GhyathSAhmed) [inline solutions/sum_numbers.py:14]: the function doc string describes the strategy it must only describes the behavior of the function

File: solutions/sum_numbers.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+Base case: n == 1 
+    return 1
+
+recursion case: for n > 1
+    smaller_result of n + the value of sum_numbers(n - 1)
+
+parameters: n = any positive integer less than 1000 because python limitation in recursion 
+
+return : the sum of all numbers less than n inclusive and greater than or equal one
```
Comment (GhyathSAhmed) [inline solutions/sum_numbers.py:24]: add the type of the returned value

File: solutions/sum_numbers.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+>>> sum_numbers(0)
+Traceback (most recent call last):
+...
+Exception: Input must be greater than 0
+"""
+
+
+    #checks n is less than 1000
+    if (n>1000):
+        raise RecursionError
```
Comment (GhyathSAhmed) [inline solutions/sum_numbers.py:43]: this should be a defensive assertion with a raise string
assert n<1000 , "Recursion Error"

File: solutions/sum_numbers.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    assert isinstance(n, int), "input must be integer"
+    
+    #checks if the value is greater than Zero
+    if n<=0:
+        raise Exception("Input must be greater than 0")
```
Comment (GhyathSAhmed) [inline solutions/sum_numbers.py:48]: - the same as the upper comment it should be an assertion.
- the message does not align with the logic, the logic is greater or equal , and the message is only greater.

File: solutions/sum_numbers.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+Examples:
+
+>>> sum_numbers(4)
+10
+
+>>> sum_numbers(5)
+15
+
+>>> sum_numbers(0)
+Traceback (most recent call last):
```
Comment (GhyathSAhmed) [inline solutions/sum_numbers.py:35]: traceback is an error because of the logic when you pass 0 it will raise exception duo to lines 47 and 48

File: solutions/sum_numbers.py (hunk 3 above)
Comment (Maab232001) [reply solutions/sum_numbers.py:43]: Thanks for your important instructions 

I didn't understand what you meant exactly here.
 Why do you suggest an assertion error while the error type is already specified as a recursion error?

File: solutions/sum_numbers.py (hunk 4 above)
Comment (Maab232001) [reply solutions/sum_numbers.py:48]: Oops!
I am gonna edit the message but I still didn't get the meaning of changing the error type

File: solutions/tests/test_sum_numbers.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    
+    def test_large_number(self):
+      """This test checks the function's performance for a large number close to recursion limit."""
+      self.assertEqual(sum_numbers(900), 405450)  # This should pass normally.
+
+    
+    # Defensive test
+    
+    def test_zero(self):
+      """This test checks if the function raises an Exception for zero input."""
```
Comment (GhyathSAhmed) [inline solutions/tests/test_sum_numbers.py:45]: it bring a message with "raise Exception: Input must be greater than 0"

File: solutions/sum_numbers.py (hunk 5 above)
Comment (Maab232001) [reply solutions/sum_numbers.py:35]: this way of writing the test was the only way for the doc-test to work so I didn't get the logic behind it

File: solutions/sum_numbers.py (hunk 3 above)
Comment (GhyathSAhmed) [reply solutions/sum_numbers.py:43]: using assertion is to verify conditions that should never fail in normal operation like for example n should never be larger than 1000.
it is a better implementation for debugging and development than using if statements

File: solutions/sum_numbers.py (hunk 4 above)
Comment (GhyathSAhmed) [reply solutions/sum_numbers.py:48]: it is not the error type that you should change it is the implementation of assertions instead of if statements

Comment (Maab232001) [reply solutions/sum_numbers.py:48]: can you help me with the code syntax I don't think that I can understand the difference

File: solutions/sum_numbers.py (hunk 3 above)
Comment (Maab232001) [reply solutions/sum_numbers.py:43]: Ok I get it now

File: solutions/sum_numbers.py (hunk 4 above)
Comment (GhyathSAhmed) [reply solutions/sum_numbers.py:48]: yes for sure the syntax is:
assert n>0, "Input must be greater than 0"

**Review by GhyathSAhmed — APPROVED** [review]
great job !!
